src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -0,0 +1,319 @@
+											scankey /*=scankey*/,
+											1 /*=num_keys*/,
+											cagg_watermark_scan_update /*=tuple_found*/,
+											RowExclusiveLock /*=lockmode*/,
+											CONTINUOUS_AGGS_WATERMARK_TABLE_NAME /*=table_name*/,
+											&new_watermark /*=data*/);
+
+	if (!watermark_updated)
+	{
+		Catalog *catalog = ts_catalog_get();
+		Relation rel =
+			table_open(catalog_get_table_id(catalog, CONTINUOUS_AGGS_WATERMARK), RowExclusiveLock);
+		TupleDesc desc = RelationGetDescr(rel);
+		Datum values[Natts_continuous_aggs_watermark];
+		bool nulls[Natts_continuous_aggs_watermark] = { false };
+		CatalogSecurityContext sec_ctx;
+
+		values[AttrNumberGetAttrOffset(Anum_continuous_aggs_watermark_mat_hypertable_id)] =
+			Int32GetDatum(mat_hypertable_id);
+		values[AttrNumberGetAttrOffset(Anum_continuous_aggs_watermark_watermark)] =
+			Int64GetDatum(new_watermark);
+
+		ts_catalog_database_info_become_owner(ts_catalog_database_info_get(), &sec_ctx);
+		ts_catalog_insert_values(rel, desc, values, nulls);
+		ts_catalog_restore_user(&sec_ctx);
+
+		table_close(rel, RowExclusiveLock);
+	}
+}
+
+TSDLLEXPORT void
+ts_cagg_watermark_set(Hypertable *mat_ht, int64 watermark, bool watermark_isnull)
+{
+	ContinuousAgg *cagg = ts_continuous_agg_find_by_mat_hypertable_id(mat_ht->fd.id);
+	const Dimension *dim = hyperspace_get_open_dimension(mat_ht->space, 0);
+
+	if (NULL == cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid materialized hypertable ID: %d", mat_ht->fd.id)));
+
+	if (watermark_isnull)
+	{
+		watermark = ts_time_get_min(ts_dimension_get_partition_type(dim));
+	}
+	else
+	{
+		Oid timetype = ts_dimension_get_partition_type(dim);
+
+		/*
+		 * The materialized hypertable is already bucketed, which means the
+		 * max is the start of the last bucket. Add one bucket to move to the
+		 * point where the materialized data ends.
+		 */
+		if (ts_continuous_agg_bucket_width_variable(cagg))
+		{
+			/*
+			 * Since `value` is already bucketed, `bucketed = true` flag can
+			 * be added to ts_compute_beginning_of_the_next_bucket_variable() as
+			 * an optimization, if necessary.
+			 */
+			watermark =
+				ts_compute_beginning_of_the_next_bucket_variable(watermark, cagg->bucket_function);
+		}
+		else
+		{
+			watermark =
+				ts_time_saturating_add(watermark, ts_continuous_agg_bucket_width(cagg), timetype);
+		}
+	}
+
+	cagg_watermark_set_internal(mat_ht->fd.id, watermark);
+
+	return;
+}
+
+TSDLLEXPORT bool
+ts_cagg_watermark_delete_by_mat_hypertable_id(int32 mat_hypertable_id)
+{
+	int count = 0;
+	ScanIterator iterator =
+		ts_scan_iterator_create(CONTINUOUS_AGGS_WATERMARK, RowExclusiveLock, CurrentMemoryContext);
+
+	cagg_watermark_init_scan_by_mat_hypertable_id(&iterator, mat_hypertable_id);
+
+	ts_scanner_foreach(&iterator)
+	{
+		TupleInfo *ti = ts_scan_iterator_tuple_info(&iterator);
+		ts_catalog_delete_tid(ti->scanrel, ts_scanner_get_tuple_tid(ti));
+		count++;
+	}
+	return count > 0;
```

Comment:
Nit: This return value is never checked. Seems redundant.

Reply:
Removed.